### DISCUSSION TITLE
BLE MIDI usin a HM-10 module

### DISCUSSION THREAD

**Discussion by KontinuumLab** [FortySevenEffects/arduino_midi_library#258] [Q&A] [answered]
I'm trying to make a BLE MIDI device using an Arduino Pro Micro and a HM-10 module.
I've looked at the BLEMIDI transport layer from lathoub, but it works with ESP32 and the Arduino Nano BLE boards. 
**_I think it would be awesome to add BLE-MIDI to the 32u4 boards, specifically because they also do MIDI over USB, so it would be cheap and easy to make a device that could do both._** 
As much as I would love to try to adapt/expand the library for my use case, my attempts so far have just left me scratching my head.

I have the HM-10 sending simple serial communication to my phone (Using the "Serial Bluetooth Terminal" app), so the hardware is working fine. Using this sketch: 
```
void setup()
{

  Serial.begin(9600);
  Serial1.begin(9600);
}
void loop()
{
  if (Serial1.available())
  {
    char i = Serial1.read();
    Serial.print(i);
    Serial1.print("Received");
  }
}
```

Using a simple Serial MIDI sketch, I'm also able to send bytes and receive them on my phone, but of course it doesn't register as a MIDI device. Using this sketch:
```
#include <MIDI.h>

// Create and bind the MIDI interface to the default hardware Serial port
MIDI_CREATE_DEFAULT_INSTANCE();

void setup()
{
    MIDI.begin(MIDI_CHANNEL_OMNI);  // Listen to all incoming messages
}
void loop()
{
    // Send note 42 with velocity 127 on channel 1
    MIDI.sendNoteOn(42, 127, 1);

    // Read incoming messages
    MIDI.read();
}
```
Does anyone know of an existing way to do this, or have the skills/desire to expand the existing library with 32u4 functionality?
Any help would be much appreciated.

**Comment by lathoub** [comment, marked as answer]
The HM-10 is a hardware rapper around the TI CC2540 chip, that has AFAIK 1 fixed service and 1 fixed characteristic (no BLE-MIDI) - so you can't have the HM-10 to announce itself as a true BLE-MIDI device (but that you knew already).

> expand the existing library with 32u4 functionality

The ATMega 32u4 has no BLE-MIDI capabilities, and it will always depend on additional hardware to be able to support BLE-MIDI (eg Adafruit Feather 32u4 Bluefruit LE). So I do not understand your question to _expand the existing library with 32u4 functionality_.

Have you looked at [Hairless MIDI](https://projectgus.github.io/hairless-midiserial/)?

**Comment by KontinuumLab** [comment]
Thanks for the swift answer!

> you can't have the HM-10 to announce itself as a true BLE-MIDI device (but that you knew already). 

I only knew that _I_ hadn't been able to do it, but was hoping that it was possible. Alas, you cleared that up for me.

> The ATMega 32u4 -/- will always depend on additional hardware to be able to support BLE-MIDI (eg Adafruit Feather 32u4 Bluefruit LE)

I was simply hoping that the HM-10 could be that additional hardware, with the right changes to the MIDI library. 
From what I understand now, my only real option for a device that can do both USB MIDI and BLE MIDI would be the Feather, like you mention. It's not exactly cheap though...  
Do I understand correctly that the USB MIDI support on that is still a work in progress? (From here: https://github.com/lathoub/Arduino-USBMIDI)
Another possible option would be the Arduino Nano 33 BLE, but has the same problem.
Wish I could help you with it, but I'm afraid most of this deep library work goes well above my head.
The ESP32 is great of course, but doesn't have native USB.

> Have you looked at Hairless MIDI?

Yes, but I'm specifically looking for a good way to connect my cheap DIY MIDI controllers to my _phone_, so that's not going to help.

Oh, well.
Thanks again for your help with this, and thanks for all the work you do.
